packages/scm/src/browser/dirty-diff/diff-computer.ts

```diff
@@ -121,9 +158,88 @@ export interface DirtyDiff {
      * Lines modified by comparison to previous revision.
      */
     readonly modified: LineRange[];
+    /**
+     * Range mappings for the diff, if {@link DirtyDiffOptions.rangeMappings requested}.
+     */
+    readonly rangeMappings?: ChangeRangeMapping[];
 }
 
+/**
+ * Represents a range that starts at the beginning of the {@link start} line
+ * and spans up to the end of the {@link end} line.
+ */
 export interface LineRange {
     start: number;
     end: number;
 }
+
+/**
+ * Represents a range that starts and ends either at the beginning of the {@link start} line or at the end of the {@link end} line.
+ */
+export type EmptyLineRange = { start: number; end?: undefined; } | { start?: undefined; end: number };
```

Comment:
Would not that break the current semantics of the `LineRange`? Currently, a `LineRange` where `start === end` represents a single line, not an empty line range.

See also https://github.com/eclipse-theia/theia/pull/13104#discussion_r1511333972.

Reply:
There is no promise in Theia that we won't break API. We try to be adopter-friendly in that we try to minimize breakage, but if keeping the API stable means we're making the code worse, change the API! (As a veteran of the Eclipse IDE project, I vividly recall the horrors of backward compatibility ;-)) 
In this particular case, there are a couple of factors I would consider:

1. I don't think that `LineRange' is something many adopters would rely on: it's more of an internal thing.
2. Adopting the changed semantics is straightforward
3. It probably makes the client code simpler as we don't have to deal with `undefined`: we can try this out by observing our internal client code.
4. Since the semantics were not documented before, you could argue that they're not part of the API

Assuming there's a "breaking change" notice in the `Readme.md`, I would have no problem changing this.

Reply:
Addressed this and [other](https://github.com/eclipse-theia/theia/pull/13104#discussion_r1495948104) concerns in https://github.com/eclipse-theia/theia/pull/13104/commits/f8c797680c437f113c38756de97f23150913daca.

I tend to agree with your arguments in this particular case, although adopting the changed semantics in `LineRange` might not be entirely "straightforward" in my experience. But let's hope that this is actually not "something many adopters would rely on".

Anyway, having lived in the Eclipse IDE land for about 20 years myself, I have come to appreciate strict rules for API evolution along with clean separation between public and internal API as one of the main strengths of Eclipse as a Platform.

I also recall a few cases where breaking even the internal API in the Eclipse Platform caused real issues downstream, including some of the Eclipse projects themselves, which forced those breaking changes to be reverted...

Do I understand it correctly that Theia currently doesn't follow such strict API policy as the Eclipse Platform?

Reply:
Theia has two levels of extensibility: compile-time extensibility via Theia extensions and runtime-extensibility via Theia plugins (aka VS Code extensions). For runtime extensibility, we have adopted the VS Code extension API, which is stable and backward-compatible. For compile-time extensibility, we have basically no API stability guarantees. 
Since the user cannot add Theia extensions to an existing Theia-based product, this does not create problems for the end-user. Any adopter that publishes a Theia-based product will have to adapt his Theia extensions if he chooses to adopt a newer release with breaking changes. That's why we have a "breaking changes" section in the changelog and adopters are encouraged to collaborate on `docs/Migration.md` in more difficult migrations.
Personally, I find this a good compromise between ease of adoption and the dreaded ISomething1...ISomething27. Want stability => VSX API. Want total flexilibity => Theia extension. Or any mixture thereof.

Reply:
I see. Thank you for the detailed explanation.